docs/madr/decisions/011-rate-limit-policy.md

```diff
@@ -0,0 +1,358 @@
+- Add support for TCP connection limiting at the basic level
+- Possibility to add global rate limiting in the future
+ 
+## Considered Options
+ 
+### Naming
+ 
+- **MeshRateLimit**
+- MeshHTTPRateLimit and MeshTCPRateLimit
+ 
+## Solution
+ 
+### Current configuration
+Below is a sample rate-limiting configuration.
+ 
+```yaml
+spec:
+ sources:
+   - match:
+       kuma.io/service: "redis_kuma-demo_svc_6379"   
+ destinations:
+   - match:
+       kuma.io/service: "demo-app_kuma-demo_svc_5000"
+ conf:
+   http:
+     requests: 5
+     interval: 10s
+     onRateLimit:
+       status: 423
+       headers:
+         - key: "x-Kuma-rate-limited"
+           value: "true"
+           append: true
+```
+We are using [Envoy's local rate limit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) which supports only HTTP configuration.
+ 
+The configuration translates to the Envoy configuration on the route and the listener filter:
+ 
+```yaml
+{
+ "match": {
+   "prefix": "/",
+   "headers": [
+   {
+     "name": "x-kuma-tags",
+     "safe_regex_match": {
+     "google_re2": {},
+     "regex": ".*&kuma.io/service=[^&]*redis_kuma-demo_svc_6379[,&].*"
+     }
+   }
+   ]
+ },
+ "route": {
+   "cluster": "localhost:5000",
+   "timeout": "0s"
+ },
+ "typed_per_filter_config": {
+   "envoy.filters.http.local_ratelimit": {
+   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
+   ...
+   }
+ }
+}
+ 
+```
+ 
+### Specification
+ 
+The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.
+Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.
+ 
+#### **HTTP Rate Limit**
+ 
+Rate limiting can be configured on both HTTP connection managers and routes. Thanks to the header `x-kuma-tags`, which is propagated and set by Envoy, we can recognize the origin of the HTTP request.
+ 
+#### Top level
+ 
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+ name: ...
+```
+Rate limiting is an inbound policy so only `from` should be configured.
+ 
+#### From level
+ 
+```yaml
+from:
+ - targetRef:
+     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+     name: ...
+```
+ 
+Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there is any route that a request originates from).
+ 
+#### Configuration
+ 
+```yaml
+default:
+  local:
+    http:
+      requests: 5
+      interval: 10s
+      onRateLimit:
+        status: 423
+        headers:
+          - key: "x-kuma-rate-limited"
+            value: "true"
+            append: true
```

Comment:
Is there a default for `onRateLimit` ? If yes could we have `x-kuma-rate-limited` but set the value to the policy name?

Reply:
I don't know...
1) We can leak policy names to the outside world
2) What do we put if we merge config from multiple policies?

Reply:
We do not create a default configuration for `onRateLimit`, if status/headers are not defined, no headers are added, and use default status set by Envoy: `429`

Reply:
Ok let's stick with this defualt behaviour